### PR TITLE
Fix return types

### PR DESCRIPTION
**Pull request by astronouth7303**
* Actual algorithm for mixed-type operations
* Fix more annotations
* Tests to verify return types

There is an argument to be made for Vector/Vector operations to return the left type.

### CODE REVIEW

**Comment by nbraud** [comment]
@astronouth7303 I don't understand what's going on.

Is it making vector/vector operations that involve subclasses of `Vector2` return the subclass? If so, what's the motivation?

**Comment by astronouth7303** [comment]
Consistency of typing without every subclass needing to wrap each operation to return the right type.

Concretely, so `ppb.vector.Vector` appears and avoiding `Vector2` showing up.

**Comment by nbraud** [comment]
Oh, I didn't realise `ppb.Vector` was different from `Vector2`.

:+1: then, as it causes the bug I mentioned in mutant-game

**Comment by nbraud** [comment]
The type annotations are probably wrong, though.

**Comment by astronouth7303** [comment]
Subclasses of the thing are implied in all annotations, afaik.

So, they all return a `Vector2`. It might be a specialization of `Vector2`, but it'll be a `Vector2`.

**Comment by nbraud** [comment]
Yes, but what I mean is that code that does a vector operation, then calls a subclass method/property on the result, would work with your change, but fail to typecheck because the typechecker thinks we return a `Vector2`

**Comment by astronouth7303** [comment]
https://mypy.readthedocs.io/en/latest/kinds_of_types.html#class-types

> Any instance of a subclass is also compatible with all superclasses

So a type annotation of `Vector2` (or it's late-bound variant `'Vector2'`) implicitly includes all subclasses of `Vector2`.

**Comment by nbraud** [comment]
@astronouth7303 OK, to clarify:

```python3
class V(Vector2):
    def method(self):
        pass

x = V(1, 0) + (0, 1)  # x's inferred type will be Vector2, not V
x.method()  # Should fail to typecheck
```

**Comment by astronouth7303** [comment]
Oh, then yes. I think this is one of the limitations of the python type checker, since (afaik), you can't have a type variable that goes between different bits of a method's annotation.

Like you can't say "x.foo() will return whatever type x is".

**Comment by pathunstrom** [comment]
That sounds like a thing that should be discussed on mypy.

File: tests/test_typing.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
+import pytest
+
+from ppb_vector import Vector2
+
+# List of operations that (Vector2, Vector2) -> Vector2
```
Comment (pathunstrom) [inline tests/test_typing.py]: This suggest that the variable names could be more clear. It took me a few moments to realize the name meant "Binary operations."

File: tests/test_typing.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+@pytest.mark.parametrize('op', BINOPS)
+def test_binop_subclass(op):
+    class V1(Vector2): pass
+    class V2(V1): pass
+
+    # Normalize for reflect
+    a = op(V1(1, 2), V2(3, 4).normalize())
+    b = op(V2(1, 2), V1(3, 4).normalize())
+    assert isinstance(a, V2)
```
Comment (pathunstrom) [inline tests/test_typing.py]: Do you mean specifically that we're returning the subclass in all cases?

File: ppb_vector/vector2.py [hunk 3]
```diff
@@ -34,6 +35,37 @@ def _mkvector(value, *, castto=_fakevector):
...
+    if len(lmro) > len(rmro):
+        return left
+    elif len(rmro) > len(lmro):
+        return right
+    else:
+        # They're equal, just arbitrarily pick one
+        return left
+
+
+def _find_lowest_vector(left: type, right: type) -> type:
```
Comment (pathunstrom) [inline ppb_vector/vector2.py]: Not actionable: I don't like the amount of private functions that this is growing into. I understand where it's coming from, thus, non-actionable.

File: ppb_vector/vector2.py [hunk 4]
```diff
@@ -34,6 +35,37 @@ def _mkvector(value, *, castto=_fakevector):
         raise ValueError(f"Cannot use {value} as a vector-like")
 
 
+@functools.lru_cache()
+def _find_lowest_type(left: type, right: type) -> type:
```
Comment (pathunstrom) [inline ppb_vector/vector2.py]: Should probably use typing.Type here.

File: tests/test_typing.py (hunk 2 above)
Comment (astronouth7303) [reply tests/test_typing.py]: Yes, because V2 is a subclass of V1, it should return the more specialized V2.

File: ppb_vector/vector2.py (hunk 4 above)
Comment (astronouth7303) [reply ppb_vector/vector2.py]: These are documented as equivalent:

> `Type[Any]` is equivalent to `Type` which in turn is equivalent to `type`, which is the root of Python’s metaclass hierarchy.

But I'll do it because metaprogramming like this can be weird.

**Comment by nbraud** [comment]
@astronouth7303 It looks like it's doable: https://gist.github.com/nbraud/a592985bb700ac74894efd8243cb08ce

**Comment by astronouth7303** [comment]
Everything mentioned so far should be taken care of.

**Comment by astronouth7303** [comment]
With annotations getting this complex, I figured we should actually run it through mypy.

It was informative.

**Comment by astronouth7303** [comment]
https://github.com/python/mypy/issues/5821 came up, and has some insights into typing.

**Comment by astronouth7303** [comment]
Sorry, this PR is getting kinda large. I feel like I've touched most of the code at this point.

**Comment by nbraud** [comment]
@astronouth7303 Yeah, there is likely no way to improve the annotations without touching all the methods.  Thanks a lot for doing it.

Small question about `Realish` (which I pronounce “relish” in my head, it's funny): I had also noticed that mypy complains about `float`s and `int`s not being `Real`s, but CPython seems to disagree:
```python3
> from numbers import Real

> isinstance(1, Real)
True

> isinstance(1.0, Real)
True
```

Is mypy in the wrong about `int`s and `float`s not being `Real`s?

**Comment by nbraud** [comment]
Re: the PR becoming too large, I guess we can split off the actual behaviour change (in selecting the return type) and the improvement in type annotations.  What do you think?

**Comment by astronouth7303** [comment]
I suppose I can. I'm not sure it helps much?

**Comment by astronouth7303** [comment]
> Small question about Realish (which I pronounce “relish” in my head, it's funny): I had also noticed that mypy complains about floats and ints not being Reals, but CPython seems to disagree:

This is because mypy doesn't pickup on [runtime registered ABC subclasses](https://docs.python.org/3/library/abc.html#abc.ABCMeta.register).

So, mypy is correct in that `numbers.Real` is not in the MRO of `int` and `float`. But, if you ask if `int` is a subclass of `Real` at runtime, the answer will be yes (because it's explicitly registered as such).

**Comment by nbraud** [comment]
@astronouth7303 Sounds like a bug that should be reported against mypy, then?
In any case, we should definitely keep the `Realish` union for now, but if the `mypy` devs fix it, we can in the future remove it.

**Comment by astronouth7303** [comment]
That's Real Complicated, and there's been Very Large Discussions about mypy and ABCs and The Number Tower (PEP 3141).

File: .travis.yml [hunk 5]
```diff
@@ -36,3 +36,4 @@ install:
   
 script:
   - pytest
+  - mypy ppb_vector tests
```
Comment (nbraud) [inline .travis.yml]: I think you need to set `MYPYPATH` to be `${VIRTUAL_ENV}`, otherwise `mypy` will complain about being unable to resolve the `hypothesis` imports.

File: ppb_vector/vector2.py [hunk 6]
```diff
@@ -1,97 +1,174 @@
...
+    y: float
+    _length: float
 
-    def __init__(self, x: Real, y: Real):
-        self.x = x
-        self.y = y
-        self.length = hypot(x, y)
+    def __init__(self, x: typing.SupportsFloat, y: typing.SupportsFloat):
+        try:
+            self.x = x.__float__()
```
Comment (nbraud) [inline ppb_vector/vector2.py]: Why is this calling the dunder, rather than using `float()` ?

File: ppb_vector/vector2.py [hunk 7]
```diff
@@ -1,97 +1,174 @@
...
+
+# Anything convertable to a Vector, including lists, tuples, and dicts
 VectorLike = typing.Union[
-    'Vector2',
-    typing.List[Real],  # TODO: Length 2
-    typing.Tuple[Real, Real],
-    typing.Dict[str, Real],  # TODO: Length 2, keys 'x', 'y'
+    'Vector2',  # Or subclasses, unconnected to the VectorOrSub typevar above
+    typing.List[typing.SupportsFloat],  # TODO: Length 2
+    typing.Tuple[typing.SupportsFloat, typing.SupportsFloat],
```
Comment (nbraud) [inline ppb_vector/vector2.py]: Replace `List` and `Tuple` with `typing.Sequence` ?

File: ppb_vector/vector2.py [hunk 8]
```diff
@@ -1,97 +1,174 @@
...
         """
-        return _mkvector(value, castto=type(cls))
+        # Use Vector2.convert() instead of type(self).convert() so that 
+        # _find_lowest_vector() can resolve things well.
+        if isinstance(value, cls):
+            return value
+        elif isinstance(value, Vector2):
+            return cls(value.x, value.y)
+        # FIXME: Allow all types of sequences
+        elif isinstance(value, (list, tuple)) and len(value) == 2:
```
Comment (nbraud) [inline ppb_vector/vector2.py]: `isinstance(value, typing.Sequence)` ?

File: ppb_vector/vector2.py [hunk 9]
```diff
@@ -1,97 +1,174 @@
...
+            return cls(value[0].__float__(), value[1].__float__())
+        # FIXME: Allow all types of mappings
+        elif isinstance(value, dict) and 'x' in value and 'y' in value and len(value) == 2:
```
Comment (nbraud) [inline ppb_vector/vector2.py]: `isinstance(value, typing.Mapping)`?

Also, do we want to enforce `len(value) == 2`?
It might be useful to be able to build a vector from a dict like `{'x': 0, 'y': 1, 'color': 'brown'}`.

File: ppb_vector/vector2.py [hunk 10]
```diff
@@ -1,97 +1,174 @@
...
+            return cls(value['x'].__float__(), value['y'].__float__())
+        else:
+            raise ValueError(f"Cannot use {value} as a vector-like")
+
+    @property
+    def length(self) -> float:
+        if not hasattr(self, '_length'):
```
Comment (nbraud) [inline ppb_vector/vector2.py]: Is the reflection dance with `hasattr` to memoize the vector length really faster than recomputing it every time?

File: ppb_vector/vector2.py [hunk 11]
```diff
@@ -1,97 +1,174 @@
...
 
-    def __sub__(self, other: VectorLike) -> 'Vector2':
+    def __sub__(self: VectorOrSub, other: VectorLike) -> VectorOrSub:
+        rtype = _find_lowest_vector(type(other), type(self))
         try:
-            other = _mkvector(other)
+            other = Vector2.convert(other)
         except ValueError:
             return NotImplemented
-        rtype = type(other) if isinstance(other, Vector2) else type(self)
```
Comment (nbraud) [inline ppb_vector/vector2.py]: Is it a good idea to move the computation of `rtype` before the `try/except`?
It can be that `_find_lowest_vector` throws an exception in cases where we would previously have returned `NotImplemented`

File: ppb_vector/vector2.py [hunk 12]
```diff
@@ -109,32 +186,32 @@ def __getitem__(self, item: typing.Union[str, int]) -> Real:
...
+    def __eq__(self: VectorOrSub, other: typing.Any) -> bool:
         if is_vector_like(other):
-            other = _mkvector(other)
+            other = Vector2.convert(other)
             return self.x == other.x and self.y == other.y
         else:
             return False
 
-    def __ne__(self, other: VectorLike) -> bool:
+    def __ne__(self: VectorOrSub, other: typing.Any) -> bool:
```
Comment (nbraud) [inline ppb_vector/vector2.py]: Can't we replace the whole function with `not self == other`?

File: tests/test_vector2_angle.py [hunk 13]
```diff
@@ -10,7 +10,7 @@
     (Vector2(1, 1), Vector2(-1, 0), 135),
     (Vector2(0, 1), Vector2(0, -1), 180),
     (Vector2(-1, -1), Vector2(1, 0), 135),
-    (Vector2(-1, -1), Vector2(-1, 0), 45),
+    (Vector2(-1, -1), Vector2(-1, 0), -45),
```
Comment (nbraud) [inline tests/test_vector2_angle.py]: Why are there unrelated changed in the angle tests?

File: tests/test_vector2_reflect.py [hunk 14]
```diff
@@ -38,7 +38,7 @@ def test_reflect_prop(initial: Vector2, normal: Vector2):
     assert initial.isclose(returned)
     note(f"initial ⋅ normal: {initial * normal}")
     note(f"reflected ⋅ normal: {reflected * normal}")
-    assert isclose((initial * normal), -(reflected * normal))
+    assert isclose((initial.dot(normal)), -(reflected.dot(normal)))
```
Comment (nbraud) [inline tests/test_vector2_reflect.py]: Why is this change needed?

File: ppb_vector/vector2.py (hunk 10 above)
Comment (nbraud) [reply ppb_vector/vector2.py]: OK, I did a quick [μbench](https://gist.github.com/nbraud/2dd0ae4904f7e97727f51722a1bba007) with `timeit`, and it looks like the speedup/slowdown of saving the result in `_length` is below the measurement noise, so I would recommend removing that code and just doing `return hypot(self.x, self.y)`.

```
$ ./dunder_bench.py
1 successive length computations
        Memoized: 0.41153700300492346
        Computed: 0.41805144600220956
        Speedup:  1.5829543758445919%
5 successive length computations
        Memoized: 1.2452478090126533
        Computed: 1.2318236450082622
        Speedup:  -1.0780315297269971%
10 successive length computations
        Memoized: 2.252152765984647
        Computed: 2.2965358549845405
        Speedup:  1.970696200996347%
25 successive length computations
        Memoized: 5.274616469978355
        Computed: 5.400348082999699
        Speedup:  2.3837110003537276%
50 successive length computations
        Memoized: 10.367006862012204
        Computed: 10.45787951600505
        Speedup:  0.8765563214376826%
100 successive length computations
        Memoized: 21.056249847984873
        Computed: 21.302848722989438
        Speedup:  1.1711433744606925%
./dunder_bench.py  81.75s user 0.00s system 99% cpu 1:21.77 total
```

**Comment by nbraud** [comment]
@astronouth7303 Thanks for the link to the mypy discussions, re: the numbers tower.

File: ppb_vector/vector2.py (hunk 6 above)
Comment (astronouth7303) [reply ppb_vector/vector2.py]: Disallowing strings.

File: .travis.yml (hunk 5 above)
Comment (astronouth7303) [reply .travis.yml]: IS THAT WHAT"S GOING ON?????????????

Comment (astronouth7303) [reply .travis.yml]: Except the build isn't using virtual environments.

File: ppb_vector/vector2.py (hunk 7 above)
Comment (astronouth7303) [reply ppb_vector/vector2.py]: The code only supports `list` and `tuple`, so ideally yes, but that's not what's implemented yet.

File: ppb_vector/vector2.py (hunk 8 above)
Comment (astronouth7303) [reply ppb_vector/vector2.py]: You'd want [`collections.abc.Sequence`](https://docs.python.org/3/library/collections.abc.html#collections.abc.Sequence)

File: ppb_vector/vector2.py (hunk 9 above)
Comment (astronouth7303) [reply ppb_vector/vector2.py]: For now, I think we want to keep exact forms to prevent mistakes.

Especially since afaik no code actually uses the dict form.

File: ppb_vector/vector2.py (hunk 11 above)
Comment (astronouth7303) [reply ppb_vector/vector2.py]: They don't throw exceptions, and they are internal functions meant for DRYing.

File: ppb_vector/vector2.py (hunk 12 above)
Comment (astronouth7303) [reply ppb_vector/vector2.py]: Probably.

File: tests/test_vector2_reflect.py (hunk 14 above)
Comment (astronouth7303) [reply tests/test_vector2_reflect.py]: Because I changed it before discovering `@override` and didn't feel like switching it back.

File: .travis.yml (hunk 5 above)
Comment (astronouth7303) [reply .travis.yml]: Yeah, this is only a thing that's happening locally.

File: ppb_vector/vector2.py (hunk 10 above)
Comment (astronouth7303) [reply ppb_vector/vector2.py]: Does computing it have any significant overhead overhead? I did this to get rid of  `_fakevector` while keeping the performance benefit of not computing the length at construction time.

File: tests/test_vector2_angle.py (hunk 13 above)
Comment (astronouth7303) [reply tests/test_vector2_angle.py]: Because passing tests.

File: ppb_vector/vector2.py (hunk 10 above)
Comment (nbraud) [reply ppb_vector/vector2.py]: @astronouth7303 That's what the microbenchmark I wrote was showing: even if you access the computed length 100 times in a row, loading it from `self._length` isn't faster than recomputing it.

File: tests/test_vector2_angle.py (hunk 13 above)
Comment (nbraud) [reply tests/test_vector2_angle.py]: Except that the tests are passing in `master`, so there is something wrong going on.

File: ppb_vector/vector2.py (hunk 9 above)
Comment (nbraud) [reply ppb_vector/vector2.py]: You added `Mapping` to the imports in a6161b9ea916fd893b4493fcf29d2b747ca1f360 but you forgot to do anything with it.

File: ppb_vector/vector2.py (hunk 6 above)
Comment (nbraud) [reply ppb_vector/vector2.py]: OK, sounds good <3

File: tests/test_vector2_reflect.py (hunk 14 above)
Comment (nbraud) [reply tests/test_vector2_reflect.py]: Yet the code right above it still uses `*`.
Please switch back to `*` for consistency (and also because it would be hella weird not to use the preferred notation in tests)

File: ppb_vector/vector2.py (hunk 8 above)
Comment (astronouth7303) [reply ppb_vector/vector2.py]: `typing.Sequence` is basically a typing-related data structure, only used to construct annotation data.

`collections.abc.Sequence` is an actual abstract class with default implementations of a pile of methods.

File: tests/test_vector2_angle.py (hunk 13 above)
Comment (astronouth7303) [reply tests/test_vector2_angle.py]: You're going to make my life difficult, I know it.

File: ppb_vector/vector2.py (hunk 9 above)
Comment (astronouth7303) [reply ppb_vector/vector2.py]: I did, this comment is outdated.

File: ppb_vector/vector2.py (hunk 10 above)
Comment (astronouth7303) [reply ppb_vector/vector2.py]: I simplified the implementation of `length`.

File: tests/test_vector2_angle.py (hunk 13 above)
Comment (astronouth7303) [reply tests/test_vector2_angle.py]: *OK.*

This is actually a bug mypy found.

The two tests in this file had the same name, so only the latter one was being run.

This exposed problems in the earlier test because it wasn't running before.

This can be moved to its own PR if y'all would really rather.
